Leaflet control

hansthen opened the pull request:
Based on the original draft by @Conengmo 

1. Create a new base class `Control`, which can be used to more easily create new control based plugins.
2. Update the original `CustomControl` so that it also accepts a `style` parameter.
3. Change all the plugins that create `L.Control` subclasses to use `Control` as a base class.

This is to some extent a work-in-progress. Once https://github.com/python-visualization/folium/pull/1898 is merged I can make `Control` inherit from `JSCSSMixin` and remove that class from all the plugins and just use the inherited method from `JSCSSMixin` inheritance to add stylesheets and javascript.

Furthermore, many of the plugins could be simplified. In the current implementation `options` are very often coded in the `_Template`. By using the `Control` base class and `JsCode` parameters as part of the specified `options` we can avoid that. Or, at least, we can move flow code from the Jinja templates into the `__init__` method, which is cleaner.

Also, I tried to make `map.LayerControl` as subclass of `Control`, but that resulted in circular dependencies. 

I did not do everything at once, because I want to take small steps and check if everything stays backward compatible.

Conengmo reviewed (changes requested):
This PR contains three separate ideas that I think we should handle separately.

CustomControl is useful I think, and something requested by users. It inherits from Control, but doesn't use anything from it, so it might as well be separate.

I'm not sure what Control is supposed to do. It's a base class for many other classes in this PR, but none of the now child classes use anything from it. No arguments are passed in `super().__init__()` and the template is overwritten. So it might as well not be there. In which case is shouldn't. Long chains of object inheritance make things difficult to grok.

Then there are also some changes in Realtime that on first glance are not linked to the control change.

So my suggestion would be to take the first and third point and deal with those in separate PRs. And then have a discussion about the second point, how the class structure of Folium should be.

hansthen commented:
Thanks for the review. I will rework it as requested.

hansthen commented:
@Conengmo I removed the parts that could be handled as separate Pull Requests. I also rewrote the `MousePosition` control plugin. I think it clarifies why adding Control allows us to simplify a large part of the code base. Let me know what you think.

I would prefer to rewrite the other plugins incrementally. However, that means that for a while there will be inconsistent approaches in the code.

Conengmo reviewed (commented):
The `MousePosition` plugin is indeed a good motivation for the new `Control` class.

The `Control` class does two things:
- it's a template for creating any Leaflet control class
- it's used in the inheritance tree to indicate something is a control.

What I find difficult is that in that second case the template and the init of `Control` should be skipped. But the init it ran, and creates some `self.options` that are not used. That sounds like it could cause issues, running extra code in the tree that's not needed.

I'd appreciate it if you could keep the PR's minimal, these style changes with all these trailing comma's make it hard to review. If you want to make style changes, please do these in a separate PR.

hansthen commented on folium/plugins/beautify_icon.py:
```diff
@@ -95,7 +95,7 @@ def __init__(
         spin=False,
         number=None,
-        **kwargs
+        **kwargs,
```
They were added by `ruff format`. I agree they do not make sense in this case. As far as I know there is no option to remove these trailing commas. I can manually undo these kind of changes, but its a hassle. I'd prefer to just blindly follow `ruff`.

Conengmo replied:
We don't run `ruff format` on this repo yet though. We use Black to autoformat and have Ruff as linter. So it's better not to run `ruff format`, since it will create a lot of changes.

Conengmo replied:
We could switch over to `ruff format` though and ditch Black. Maybe that's good to do anyway, and if we merge that before this PR that also solves the issue.

hansthen replied:
My bad, I thought we used ruff as a formatter as well. I can create an extra PR to ditch black if you want, but I have no strong opinion either way. If not I will undo the ruff formatting changes.

hansthen commented on folium/plugins/realtime.py:
```diff
@@ -42,7 +42,7 @@ class Realtime(JSCSSMixin, MacroElement):
         Should missing features between updates been automatically
         removed from the layer
-    container: Layer, default GeoJson
+    container: FeatureGroup, default GeoJson
```
Will rebase once the other PR is accepted.

hansthen commented on folium/plugins/timestamped_geo_json.py:
```diff
@@ -1,14 +1,14 @@
 
 
-class TimestampedGeoJson(JSCSSMixin, MacroElement):
+class TimestampedGeoJson(JSCSSMixin, Control):
```
I agree with you that it makes it confusing here, but my remedy would go the other direction.

My feeling here is that adding both a `Layer` and a `Control` is a design mistake. There are tickets open in which users have requested to be able to add multiple `Layers` and control them using a single control. Like the `Timeline` plugin does. If I find the courage I would split the different `Timestamped` into seperate `Control` and `Layer` implementations (or rewrite them using the `TimelineControl` plugin).

Conengmo replied:
That's fair, that they should be separate. But until we rebuild this we have the current situation, where the object is not a control or a layer, but a Folium specific something.

hansthen replied:
Okay, I will remove the `Control` base class for these classes for now.

hansthen commented on folium/features.py:
```diff
@@ -1973,3 +1985,65 @@ def __init__(
             self.add_child(PolyLine(val, color=key, weight=weight, opacity=opacity))
+
+
+class Control(MacroElement):
```
Good point and the answer is I don't know yet. 

I noted that the `CustomControl` is like a `Control` that is linked to a layer, something I did not realize at first. In pure `Leaflet` this is something of an anomaly, as both `Control`s and `Layer`s are contained in a `Map` and there is no structural association between `Control` and `Layer`. Typically it is the other way round and a `Control` determines the visibility of the `Layer`.

My guess is it would be fine as you are implementing it now. It would probably be the same as other `Controls` that are associated with a set of `Layers`.

Conengmo replied:
> I noted that the CustomControl is like a Control that is linked to a layer, something I did not realize at first. In pure Leaflet this is something of an anomaly, as both Controls and Layers are contained in a Map and there is no structural association between Control and Layer. Typically it is the other way round and a Control determines the visibility of the Layer.

That's interesting... So I guess the custom control is a bit of a weird one, and it's okay not to consider it here.

hansthen commented on folium/plugins/mouse_position.py:
```diff
@@ -83,19 +53,25 @@ def __init__(
+            lng_formatter=JsCode(lng_formatter) if lng_formatter else None,
+            **kwargs,
+        )
+        self.add_js_link(
+            "Control_MousePosition_js",
+            "https://cdn.jsdelivr.net/gh/ardhi/Leaflet.MousePosition/src/L.Control.MousePosition.min.js",
+        )
+        self.add_css_link(
+            "Control_MousePosition_css",
+            "https://cdn.jsdelivr.net/gh/ardhi/Leaflet.MousePosition/src/L.Control.MousePosition.min.css",
```
Why? The method calls seem cleaner to me compared to the static attributes?

Conengmo replied:
the static attributes allow users that want to create offline maps to bundle these dependencies and then update the values in the static attributes. I know they might also use these new methods, but it's breaking behavior to define these in the init for this use case.

hansthen commented:
First off, thanks for looking at the Pull Request. I know you are not enthusiastic about making major changes to Folium, so it is appreciated that you are willing to consider this. 

> it's used in the inheritance tree to indicate something is a control.
Indeed. I also have the vague hope it will in the future help the developers of `streamlit-folium`. This package allows for dynamic changes to the generated Leaflet map. They do fairly complicated stuff now with regular expressions to separate `Layer`s and `Control`s. Having a simple marker in the python object hierarchy would make their life simpler.

> What I find difficult is that in that second case the template and the init of `Control` should be skipped. But the init it ran, and creates some `self.options` that are not used. That sounds like it could cause issues, running extra code in the tree that's not needed.

Just to clarify, you have no difficulties with how I rewrote the `mouse_position`? It is rather that in all the other instances, there is an unused parent class? I agree with you there, that the situation is not ideal. I would rewrite all the plugins (and native Controls) similarly, but it is a lot of work. Not all of it will be as straightforward as for the `mouse_position` plugin. I would prefer to approach this incrementally after (and if) we come to an agreement on the basic approach.

Would it help if I added the `Control` class also incrementally whenever I update the a plugin to use the new approach?

Conengmo commented:
> I also have the vague hope it will in the future help the developers of streamlit-folium. This package allows for dynamic changes to the generated Leaflet map. They do fairly complicated stuff now with regular expressions to separate Layers and Controls. Having a simple marker in the python object hierarchy would make their life simpler.

That's interesting! I'd like to know more about that.

I gotta run, but just as a quick thought for now. Does it make sense to split this class into a `Control` class that's basically empty but indicates the object name in Javascript is a control, and a `ControlTemplate` or whatever (bad name) that can be used to create any control object? Where `Control` is inherited from by whichever class it applies to, and `ControlTemplate` inherits from `Control` and is used to build control objects?

I understand a bit better know to have our Python classes only implement one Javascript class. Question I have now is how we then bundle these in a way that's convenient for our Python users?

Conengmo commented:
> Just to clarify, you have no difficulties with how I rewrote the mouse_position?

No, I like the abstraction. Only doubt would be whether the abstraction is common enough to warrant abstracting it.

> Would it help if I added the Control class also incrementally whenever I update the a plugin to use the new approach?

Not sure if I fully understand, but if you're talking about not adding Control as a parent to a class that implements multiple Javascript classes, then yes, doing that incrementally sounds like a good plan.

hansthen commented:
> Not sure if I fully understand, but if you're talking about not adding Control as a parent to a class that implements multiple Javascript classes, then yes, doing that incrementally sounds like a good plan.

I meant that in the current PR, I added `Control` already to several python classes that generate a Leaflet Control without adapting their implementation. These classes just have a different base class, but do not use their parent's template of constructor. This should not break anything, but it is maybe not the cleanest solution. I wanted to rewrite all the plugins later to cleanup the code, calling the parent constructor and probably removing their template code. 

My alternative would be to not add `Control` as base class for the other plugins now, but only after I have adapted the code.
